### PR TITLE
[feature] shared bus added

### PR DESCRIPTION
**Pull request by m4github**
The sharedBus feature added in this branch.
To see the "how to use" and test go to https://github.com/DoneTech-IoT/InterCom-test-code repo

### CODE REVIEW

**Comment by Behn4m** [comment]
There is no description for the PR! you missed [feature] at the beginning of the PR name. 
@m4github

**Comment by m4github** [comment]
> There is no description for the PR! you missed [feature] at the beginning of the PR name. @m4github

fixed

File: components/Utilities/SharedBus/SharedBus.c [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+        //permission to all task to peek
+        EventBits = xEventGroupSetBits(
+                        EventGroupHandleLocal, /* The event group being updated. */
+                        BIT_22);                               
+    }
+        
+    return false;    
+}
+esp_err_t SharedBusRecieve(
```
Comment (Behn4m) [inline components/Utilities/SharedBus/SharedBus.c:55]: place a new line between function definitions

File: components/Utilities/SharedBus/SharedBus.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+typedef struct {
+    uint8_t SourceID;
+    uint8_t PacketID;    
+    char data[1024];
+} SharedBusPacket_t;
+
+esp_err_t SharedBusInit();
+
+esp_err_t SharedBusSend(SharedBusPacket_t SharedBusPacket);
```
Comment (Behn4m) [inline components/Utilities/SharedBus/SharedBus.h:35]: add doxygen comment for all functions please

File: main/ServiceManager/ServiceManager.cpp [hunk 3]
```diff
@@ -128,6 +129,10 @@ void ServiceMangerTaskInit()
  */
 void ServiceMangerInit()
 {
+    if (SharedBusInit())
```
Comment (Behn4m) [inline main/ServiceManager/ServiceManager.cpp:132]: use `{ ... }` for all `if` statements, even if is has one line of code inside

File: main/ServiceManager/ServiceManager.cpp [hunk 4]
```diff
@@ -128,6 +129,10 @@ void ServiceMangerTaskInit()
...
+        ESP_LOGE(TAG, "Failed to Initialize SharedBus.");
```
Comment (Behn4m) [inline main/ServiceManager/ServiceManager.cpp:133]: there is no `break` or `return` after the error, so your code will print: 
```
Failed to Initialize SharedBus.
Initialize SharedBus.
``` 
if the initialization fails!

File: main/ServiceManager/ServiceManager.cpp [hunk 5]
```diff
@@ -128,6 +129,10 @@ void ServiceMangerTaskInit()
...
+    ESP_LOGI(TAG, "Initialize SharedBus.");
```
Comment (Behn4m) [inline main/ServiceManager/ServiceManager.cpp:134]: should be something like: "SharedBus initialized" or "initialized SharedBus successfully"

File: components/Utilities/SharedBus/SharedBus.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+extern "C"
+{
+#endif
+
+#include <freertos/FreeRTOS.h>
+#include <freertos/queue.h>
+#include <freertos/task.h>
+#include <freertos/event_groups.h>
+
+typedef enum {
```
Comment (Behn4m) [inline components/Utilities/SharedBus/SharedBus.h:14]: I feel this enum should be somewhere in `main.c` or `main.h`

File: components/Utilities/SharedBus/SharedBus.c [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+                    EventGroupHandleLocal,   /* The event group being tested. */
+                    BIT_22 | BIT_21,         /* The bits within the event group to wait for. */
+                    pdFALSE,        /* BIT_22 should NOT be cleared before returning. */
+                    pdFALSE,         /* Wait for 22th bit, either bit will do. */
+                    1);/*           Wait a maximum of 100ms for either bit to be set. */    
+
+    if((EventBits & BIT_22)) //default state
+    {           
+        ret = true;    
+        if(xQueuePeek(QueueHandle, SharedBusPacket, 1) == pdTRUE)
```
Comment (Behn4m) [inline components/Utilities/SharedBus/SharedBus.c:70]: nested `if` alert!

**Review by Behn4m — CHANGES_REQUESTED** [review]
great work.
some minor changes are needed.

**Comment by MohammadAbbasi76** [comment]
@m4github  where is your testing result?
you have to pass all tests and after that, you can merge it.
image

**Review by MohammadAbbasi76 — CHANGES_REQUESTED** [review]
where is your testing result? please comment it or send a photo from it.

File: main/ServiceManager/ServiceManager.cpp (hunk 4 above)
Comment (m4github) [reply main/ServiceManager/ServiceManager.cpp:133]: fixed.

File: main/ServiceManager/ServiceManager.cpp (hunk 5 above)
Comment (m4github) [reply main/ServiceManager/ServiceManager.cpp:134]: fixed.

File: main/ServiceManager/ServiceManager.cpp (hunk 3 above)
Comment (m4github) [reply main/ServiceManager/ServiceManager.cpp:132]: fixed.

File: components/Utilities/SharedBus/SharedBus.c [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+#include "SharedBus.h"
+#include "Custom_Log.h"
+
+#define UI_EVENT_FLAG	( 1 << UI_INTERFACE_ID )
+#define UI_EVENT_FLAG	( 1 << UI_INTERFACE_ID )
+
+#define BIT_23	( 1 << 23 ) //prevent overwriting bus for send before one turn of the event loop
+#define BIT_22	( 1 << 22 ) //peek to receive switch flag 
+#define BIT_21	( 1 << 21 ) //peek to receive switch flag 
+
```
Comment (MohammadAbbasi76) [inline components/Utilities/SharedBus/SharedBus.c:11]: please write them in .h file

File: components/Utilities/SharedBus/SharedBus.c (hunk 1 above)
Comment (m4github) [reply components/Utilities/SharedBus/SharedBus.c:55]: done.

File: components/Utilities/SharedBus/SharedBus.h (hunk 2 above)
Comment (m4github) [reply components/Utilities/SharedBus/SharedBus.h:37]: done.

File: components/Utilities/SharedBus/SharedBus.h (hunk 6 above)
Comment (m4github) [reply components/Utilities/SharedBus/SharedBus.h:14]: For some reason, I say NO. we're going to discuss it in the daily meeting.

File: components/Utilities/SharedBus/SharedBus.c (hunk 7 above)
Comment (m4github) [reply components/Utilities/SharedBus/SharedBus.c:84]: Yes but easy to read the code.

**Comment by m4github** [comment]
> where is your testing result? please comment it or send a photo from it.

here is a whole repository for different situations and tests: https://github.com/DoneTech-IoT/InterCom-test-code

File: components/Utilities/SharedBus/SharedBus.c (hunk 8 above)
Comment (m4github) [reply components/Utilities/SharedBus/SharedBus.c:11]: they are all used locally so it's better to keep them here.

File: components/Utilities/SharedBus/SharedBus.c [hunk 9]
```diff
@@ -47,27 +53,38 @@ esp_err_t SharedBusSend(SharedBusPacket_t SharedBusPacket)
...
-                    1);/*           Wait a maximum of 100ms for either bit to be set. */    
+                    pdFALSE,        /* BIT 21 or 22 should NOT be cleared before returning. */
+                    pdFALSE,         /* Wait for 21 or 22th bit, either bit will do. */
+                    1);/*           Wait a maximum of 1ms for either bit to be set. */    
 
-    if((EventBits & BIT_22)) //default state
-    {           
-        ret = true;    
-        if(xQueuePeek(QueueHandle, SharedBusPacket, 1) == pdTRUE)
+    if((EventBits & BIT_22) == 0) //default state
```
Comment (Behn4m) [inline components/Utilities/SharedBus/SharedBus.c:81]: I'm still waiting for an update here. You have four nested if statements, which makes the code difficult to read.

File: main/ServiceManager/ServiceManager.cpp [hunk 10]
```diff
@@ -129,11 +129,17 @@ void ServiceMangerTaskInit()
  */
 void ServiceMangerInit()
 {
+    nvsFlashInit();
+
     if (SharedBusInit())
```
Comment (Behn4m) [inline main/ServiceManager/ServiceManager.cpp:134]: another wrong `if` statement here. This is how it should be written:
```
uint8_t init = SharedBusInit();
if (!init)
{
    ESP_LOGE(TAG, "Failed to Initialize SharedBus.")
}

ESP_LOGI(TAG, "initialized SharedBus successfully");

**Review by Behn4m — COMMENTED** [review]
Great work on the code! However, I still need to stick with the `if` structure we previously discussed.

**Comment by Behn4m** [comment]
> @m4github where is your testing result? you have to pass all tests and after that, you can merge it. image

This PR is not using the SharedBus in the code yet, so no test result is needed. In the next steps, where the SharedBus is going to be used to transfer data between components, full integration tests would be needed for each PR

File: main/ServiceManager/ServiceManager.cpp (hunk 10 above)
Comment (m4github) [reply main/ServiceManager/ServiceManager.cpp:134]: 1- your suggestion is wrong in the "else" state    
2- I am not happy with this style
3- what you ask for is totally the same as what I have done   
ANYWAY fixed and done

**Review by Behn4m — APPROVED** [review]

**Review by Behn4m — APPROVED** [review]
good work. thank you

**Pull request merged**